File: solutions/zig/src/year2022/day01.zig
const std = @import("std");

fn parseTotalCalories(input: []const u8, allocator: std.mem.Allocator) !std.ArrayList(u32) {
    var elves_iter = std.mem.split(u8, std.mem.trim(u8, input, " \n"), "\n\n");
    var total_calories = std.ArrayList(u32).init(allocator);

    while (elves_iter.next()) |elf| {
        var calories: u32 = 0;
        var calories_iter = std.mem.split(u8, elf, "\n");
        while (calories_iter.next()) |cal| {
            calories += try std.fmt.parseInt(u32, cal, 10);
        }
        try total_calories.append(calories);
    }

    return total_calories;
}

fn part1(calories: *std.ArrayList(u32)) u32 {
    return std.mem.max(u32, calories.items);
}

fn part2(calories: *std.ArrayList(u32)) u32 {
    var top_three = [3]u32{ 0, 0, 0 };
    for (calories.items) |cal| {
        inline for (0..top_three.len) |i| {
            if (cal > top_three[i]) {
                comptime var j: usize = 2;
                inline while (j > 0) : (j -= 1) {
                    top_three[j] = top_three[j - 1];
                }
                top_three[i] = cal;
                break;
            }
        }
    }

    return top_three[0] + top_three[1] + top_three[2];
}

pub fn solution(input: []const u8, allocator: std.mem.Allocator) !struct { []u8, []u8 } {
    var calories = try parseTotalCalories(input, allocator);
    defer calories.deinit();

    return .{
        try std.fmt.allocPrint(allocator, "{}", .{part1(&calories)}),
        try std.fmt.allocPrint(allocator, "{}", .{part2(&calories)}),
    };
}

File: solutions/go/pkg/year2022/day01.go
package year2022

import (
	"slices"
	"strconv"
	"strings"
)

func parseTotalCalories(input string) []int {
	elves := strings.Split(strings.TrimSpace(input), "\n\n")
	var totalCalories []int

	for _, elf := range elves {
		calories := 0
		for _, cal := range strings.Split(elf, "\n") {
			cal, err := strconv.Atoi(cal)
			if err != nil {
				panic(err)
			}
			calories += cal
		}
		totalCalories = append(totalCalories, calories)
	}

	return totalCalories
}

func day1Part1(calories []int) int {
	return slices.Max(calories)
}

func day1Part2(calories []int) int {
	topThree := []int{0, 0, 0}
	for _, cal := range calories {
		for i := 0; i < len(topThree); i++ {
			if cal > topThree[i] {
				for j := 2; j > 0; j-- {
					topThree[j] = topThree[j-1]
				}
				topThree[i] = cal
				break
			}
		}
	}

	return topThree[0] + topThree[1] + topThree[2]
}

func Day1(input string) (string, string) {
	calories := parseTotalCalories(input)

	return strconv.Itoa(day1Part1(calories)), strconv.Itoa(day1Part2(calories))
}
